kclvm/runner/src/lib.rs

```diff
@@ -1,2 +1,133 @@
+/// KclvmAssembler is mainly responsible for concurrent compilation of multiple files.
+/// Single-file compilation in each thread in concurrent compilation is the responsibility of KclvmLibAssembler.
+/// In the future, it may support the dynamic link library generation of multiple intermediate language.
+/// KclvmLibAssembler currently only supports LLVM IR.
+///
+/// After linking all dynamic link libraries by KclvmLinker, method "KclvmLinker::link_all_libs" will return a path
+/// for dynamic link library after linking.
+///
+/// At last, KclvmRunner will be constructed and call method "run" to execute the kcl program.
+///
+/// TODO: Need to be a better backend abstraction
+///
+/// # Examples
+///
+/// ```
+/// use kclvm_runner::{execute, runner::ExecProgramArgs};
+/// use kclvm_parser::load_program;
+/// use kclvm_ast::ast::Program;
+///
+/// // Default plugin agent
+/// let plugin_agent = 0;
+///
+/// // Get default args
+/// let args = ExecProgramArgs::default();
+/// let opts = args.get_load_program_options();
+///
+/// // Parse kcl file
+/// let kcl_path = "./src/test_datas/init_check_order_0/main.k";
+/// let prog = load_program(&[kcl_path], Some(opts)).unwrap();
+///     
+/// // Resolve ast, generate libs, link libs and execute.
+/// // Result is the kcl in json format.
+/// let result = execute(prog, plugin_agent, &args).unwrap();
+/// ```
+pub fn execute(
+    mut program: Program,
+    plugin_agent: u64,
+    args: &ExecProgramArgs,
+) -> Result<String, String> {
+    // Resolve ast
+    let scope = resolve_program(&mut program);
+    scope.check_scope_diagnostics();
+
+    // Create a temp entry file and the temp dir will be delete automatically
+    let temp_dir = tempdir().unwrap();
+    let temp_dir_path = temp_dir.path().to_str().unwrap();
+    let temp_entry_file = temp_file(temp_dir_path);
+    // Generate libs
+    let lib_paths = assembler::KclvmAssembler::default().gen_libs(
+        program,
+        scope,
+        plugin_agent,
+        &temp_entry_file,
+        KclvmLibAssembler::LLVM(LlvmLibAssembler {}),
+    );
+
+    // Link libs
+    let lib_suffix = Command::get_lib_suffix();
+    let temp_out_lib_file = format!("{}.out{}", temp_entry_file, lib_suffix);
+    let lib_path = linker::KclvmLinker::link_all_libs(lib_paths, temp_out_lib_file, plugin_agent);
+
+    // Run
+    let runner = KclvmRunner::new(
+        lib_path.as_str(),
+        Some(KclvmRunnerOptions {
+            plugin_agent_ptr: plugin_agent,
+        }),
+    );
+    let result = runner.run(args);
+
+    // Clean temp files
+    remove_file(&lib_path);
+    clean_tmp_files(&temp_entry_file, &lib_suffix);
+    result
+}
+
+/// Clean all the tmp files generated during lib generating and linking.
+#[inline]
+pub fn clean_tmp_files(temp_entry_file: &String, lib_suffix: &String) {
+    let temp_entry_lib_file = format!("{}{}", temp_entry_file, lib_suffix);
+    remove_file(&temp_entry_lib_file);
+}
+
+#[inline]
+pub fn remove_file(file: &str) {
```

Comment:
use private function `pub fn -> fn`

Reply:
I called this function in unit test, so I use pub fn -> pub(crate) fn.